### PR TITLE
Change the inner iterations logic

### PR DESCRIPTION
**Pull request by totto82**
This replace the Boolean switch to enable inner iterations with a
int that controls for which maximum number of newton iterations
inner iterations applies.

Default is set to 3

### CODE REVIEW

**Comment by totto82** [comment]
This replace #3331 where the 3 limit is hardcoded.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
This has now been tested on
1) Jenkins: Results are qualitatively the same. I have manually investigated the suspicious test failures. There are differences but they are minor and related to different time-stepping due to changes in convergence. Both Stw and MSW wells seem to behave ok with the default value 3 that I have given to the parameter
2) Benchmark: The model2 issue we have seen earlier for inner iterations is almost gone. Norne shows a slight improvement. The NORNE_4C results are inconclusive. 
3) The norne family: For STW models this gives improvements. For MSW the results are similar
4) Testing from Equinor side: Gives good speed up on tested models. Including an ensemble run. 

--> Lets merge this!

The old behavior can still be obtained by setting the parameter to 0 for no inner iterations and 99 for always inner iterations.

**Comment by bska** [comment]
> Lets merge this!

Please don't update the reference data until OPM/opm-common#2485 is merged along with its associate new reference solutions in OPM/opm-tests#550.

**Comment by totto82** [comment]
@bska Any comment on the implementation? Seems like @GitPaean is busy.

**Review by bska — APPROVED** [review]
Yes, this looks good to me.  I'll update the reference solutions once the current in-flight data update has landed.

**Comment by bska** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
sorry for not seeing the notification. I am not that keen to unify the STW and MSW in term of this perspective. Ideally, both STW and MSW should only apply converged results to the mass balance equations, but we just do not know why it degenerates the performance for STW (we kind of knows some, but not sure what to do).  And also, all these types of logic makes the code complicated and it brings complaints, but it is a really complicated business.  It is a small change that is easier to change later when necessary. I have no objection here.

**Comment by bska** [comment]
jenkins build this opm-tests=551 please

**Comment by bska** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=551 please

**Comment by bska** [comment]
This keeps failing a **parallel** regression test ([ACTIONX_M1](https://ci.opm-project.org/job/opm-simulators-PR-builder/2560/testReport/(root)/mpi/compareParallelSim_flow_ACTIONX_M1/)).  We can't merge while that happens so I think we need to revisit the logic here.

@totto82: Please have a closer look at the failing test and see if you're able to reproduce it on your local machine.

**Comment by totto82** [comment]
jenkins build this opm-tests=551 please

File: opm/simulators/flow/BlackoilModelParametersEbos.hpp [hunk 1]
```diff
@@ -405,11 +394,10 @@ namespace Opm
             relaxed_inner_tolerance_flow_ms_well_ = EWOMS_GET_PARAM(TypeTag, Scalar, RelaxedFlowTolInnerIterMsw);
             relaxed_inner_tolerance_pressure_ms_well_ = EWOMS_GET_PARAM(TypeTag, Scalar, RelaxedPressureTolInnerIterMsw);
             max_pressure_change_ms_wells_ = EWOMS_GET_PARAM(TypeTag, Scalar, MaxPressureChangeMsWells);
-            use_inner_iterations_ms_wells_ = EWOMS_GET_PARAM(TypeTag, bool, UseInnerIterationsMsWells);
             max_inner_iter_ms_wells_ = EWOMS_GET_PARAM(TypeTag, int, MaxInnerIterMsWells);
             strict_inner_iter_ms_wells_ = EWOMS_GET_PARAM(TypeTag, int, StrictInnerIterMsWells);
             regularization_factor_ms_wells_ = EWOMS_GET_PARAM(TypeTag, Scalar, RegularizationFactorMsw);
-            use_inner_iterations_wells_ = EWOMS_GET_PARAM(TypeTag, bool, UseInnerIterationsWells);
+            max_niter_inner_well_iter_ = EWOMS_GET_PARAM(TypeTag, bool, MaxNewtonIterationsWithInnerWellIterations);
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/simulators/flow/BlackoilModelParametersEbos.hpp:400]: I got the following error when running one case, 
```Processing grid
Total number of active cells: 2794 / total pore volume: 58414012 RM3
Program threw an exception: GET_*_PARAM for parameter 'MaxNewtonIterationsWithInnerWellIterations' called with at least two different types (int and bool)```

[the remark this answers is not in the document]
Comment (bska) [reply opm/simulators/flow/BlackoilModelParametersEbos.hpp:400]: > Should not here be a `int`? @totto82

Fixed in #3370.

[the remark this answers is not in the document]
Comment (GitPaean) [reply opm/simulators/flow/BlackoilModelParametersEbos.hpp:400]: okay. Thanks. I did not see that one.

**Pull request merged**